src/Model/Operation/ProductSyncHandler.php

```diff
@@ -145,18 +148,25 @@ private function doUpsertOperation(
             $channelId,
         );
 
-        /** @var SalesChannelProductEntity $product */
+        /** @var ProductEntity $product */
         foreach ($productCollection as $product) {
             // TODO: up to 2MB payload!
             $nostoProducts = [];
             foreach ($this->processProductVariants($product, $context) as $handledProduct) {
-                $nostoProducts[] = $this->handleProduct(
-                    $handledProduct,
-                    $context,
-                    $account,
-                    $hideProductsAfterClearance,
-                    $ids,
-                );
+                $shopwareProduct = $this->getShopwareProduct($handledProduct->id, $context);
+                if (!is_null($shopwareProduct)) {
+                    $shopwareProduct->setChildren($handledProduct->getChildren());
+                    $nostoProducts[] = $this->handleProduct(
+                        $shopwareProduct,
+                        $context,
+                        $account,
+                        $hideProductsAfterClearance,
+                        $ids,
+                    );
+                } else {
+                    $this->deleteVariantProducts($handledProduct, $context, $account, $ids);
+                    $this->doDeleteOperation($account, $context, [$handledProduct->getParentId()], $ids);
```

Comment:
Don't we need to delete the product itself as well? You are just deleting the parent ID here.

Reply:
The line above deletes all variants of that product, so the entire product would be deleted. Is there a better way to do this?

Reply:
Yes, the variants are deleted, but not the product itself. At least I don't see a reason why the product should be kept.

```suggestion
                    $this->doDeleteOperation(
                        $account,
                        $context,
                        [$handledProduct->getId(), $handledProduct->getParentId()],
                        $ids
                    );
```